filesystems/blob/src/main/scala/cromwell/filesystems/blob/BlobPathBuilder.scala

```diff
@@ -57,35 +57,43 @@ object BlobPathBuilder {
 
 class BlobPathBuilder(blobTokenGenerator: BlobTokenGenerator, container: String, endpoint: String) extends PathBuilder {
 
-  val credential: AzureSasCredential = new AzureSasCredential(blobTokenGenerator.getAccessToken)
-  val fileSystemConfig: Map[String, Object] = Map((AzureFileSystem.AZURE_STORAGE_SAS_TOKEN_CREDENTIAL, credential),
-                                                  (AzureFileSystem.AZURE_STORAGE_FILE_STORES, container),
-                                                  (AzureFileSystem.AZURE_STORAGE_SKIP_INITIAL_CONTAINER_CHECK, java.lang.Boolean.TRUE))
-
-  def retrieveFilesystem(uri: URI): Try[FileSystem] = {
-    Try(FileSystems.getFileSystem(uri)) recover {
-      // If no filesystem already exists, this will create a new connection, with the provided configs
-      case _: FileSystemNotFoundException => FileSystems.newFileSystem(uri, fileSystemConfig.asJava)
-    }
-  }
-
   def build(string: String): Try[BlobPath] = {
     validateBlobPath(string, container, endpoint) match {
-      case ValidBlobPath(path) => for {
-            fileSystem <- retrieveFilesystem(new URI("azb://?endpoint=" + endpoint))
-            nioPath <- Try(fileSystem.getPath(path))
-            blobPath = BlobPath(nioPath, endpoint, container)
-          } yield blobPath
+      case ValidBlobPath(path) => Try(BlobPath(path, endpoint, container, blobTokenGenerator))
       case UnparsableBlobPath(errorMessage: Throwable) => Failure(errorMessage)
     }
   }
-
   override def name: String = "Azure Blob Storage"
 }
 
-// Add args for container, storage account name
-case class BlobPath private[blob](nioPath: NioPath, endpoint: String, container: String) extends Path {
-  override protected def newPath(nioPath: NioPath): Path = BlobPath(nioPath, endpoint, container)
+object BlobPath {
+  def buildConfigMap(credential: AzureSasCredential, container: String): Map[String, Object] = {
+    Map((AzureFileSystem.AZURE_STORAGE_SAS_TOKEN_CREDENTIAL, credential),
+      (AzureFileSystem.AZURE_STORAGE_FILE_STORES, container),
+      (AzureFileSystem.AZURE_STORAGE_SKIP_INITIAL_CONTAINER_CHECK, java.lang.Boolean.TRUE))
+  }
+
+  def findNioPath(path: String, endpoint: String, container: String, blobTokenGenerator: BlobTokenGenerator): NioPath = (for {
+      fileSystem <- retrieveFilesystem(new URI("azb://?endpoint=" + endpoint), container, blobTokenGenerator)
+      nioPath <- Try(fileSystem.getPath(path))
+    } yield nioPath).get // Ideally we would unwrap this to a NioPath on success and on a access failure try to recover
```

Comment:
Running into issues here. If we can in theory know that the filesystem auth has expired in this method, we could essentially guarantee that we return a valid filesystem by trying to refresh. There are however some errors that I am not sure how we should process (like invalid path).

If we can figure out how to get this method and as a result the below nioPath method below to be the encapsulation of this refreshing behavior, we don't have to reimplement all of the filesystem methods that try to interact with the filesystem, which I think is ideal.

Reply:
So I think I figured out that using a method called checkAccess is how the AzureFileSystem provider approaches identifying if the path has the auth needed to proceed, so I used this to move the onus of the check to the findNioPath method. This meant that we can recover from the thrown exceptions when first building the path object, rather than when trying to access it, however this doesn't really address the use of exceptions for control flow which I am not sure is the best way to go.